lib/publiccloud/azure.pm

```diff
@@ -205,4 +205,56 @@ sub on_terraform_timeout {
     assert_script_run("az group delete --yes --no-wait --name $resgroup");
 }
 
+sub get_state_from_instance
+{
+    my ($self, $instance) = @_;
+    my $name = $instance->instance_id();
+    my $out = decode_json(script_output("az vm get-instance-view --name $name --resource-group $name --query instanceView.statuses[1] --output json", quiet => 1));
+    die("Expect PowerState but got " . $out->{code}) unless ($out->{code} =~ m'PowerState/(.+)$');
+    return $1;
+}
+
+sub get_ip_from_instance
+{
+    my ($self, $instance) = @_;
+    my $name = $instance->instance_id();
+
+    my $out = decode_json(script_output("az vm list-ip-addresses --name $name --resource-group $name", quiet => 1));
+    return $out->[0]->{virtualMachine}->{network}->{publicIpAddresses}->[0]->{ipAddress};
+}
+
+sub stop_instance
```

Comment:
looks like generic enough to get abstract method in `publiccloud::provider` . you can leave other providers implementations empty but when we will need such functionality for other providers we will use something common

Reply:
sorry I should read till the before writing something :) of course you did it . ignore my comment